File: [0120][Medium][Triangle]/Triangle_2.java
// leetcode time     cost : 2 ms
// leetcode memory   cost : 39.1 MB
// solution 1, DP by iterate from the last 2nd row to the top row.
import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;
import java.math.*;
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int row = triangle.size();
        int[] minlen = new int[row+1];
        for (int level = row-1;level>=0;level--){
            for (int i = 0;i<=level;i++){   //row i has (i+1)  numbers
                minlen[i] = Math.min(minlen[i], minlen[i+1]) + triangle.get(level).get(i);
                
            }
        }
        return minlen[0];
    }
}

public class Triangle_2 {
    public static void main(String args[]) {
        int[][] triangle_int = {{2},{3,4},{6,5,7},{4,1,8,3}} ;   // expect is 11
        List<List<Integer>> triangle = new ArrayList<List<Integer>>();
        
         for(int i = 0;i < triangle_int.length;i++){
            // need to new the tmp object in the loop, otherwise the triangle.add(i,tmp) will overwrite all the elements before 
            // due to the tmp reference called in the add function.
            List<Integer> tmp = new ArrayList<Integer>();
            for(int j =0;j <= i;j++){
                tmp.add(j,triangle_int[i][j]);
            }
            triangle.add(i,tmp);
            //System.out.println("i is :"+i+",triangle is :" + triangle+",tmp: "+tmp);
        } 
        Solution Solution_obj = new Solution();
        int result = Solution_obj.minimumTotal(triangle);
        System.out.println("In java code,return value is :" + (result));
    }
}

File: [0120][Medium][Triangle]/Triangle_2.py
# -*- coding: utf-8 -*-  
# leetcode time     cost : 72 ms
# leetcode memory   cost : 14.4 MB
# solution 1, DP by iterate from the last 2nd row to the top row.
# 6 coding solutions as below.
# https://leetcode.com/problems/triangle/discuss/38827/One-liner-in-Python
from functools import reduce
# reduce(function, iterable[, initializer])
class Solution:
    # def minimumTotal(self, triangle: List[List[int]]) -> int:
    def minimumTotal(self, t):
        return reduce(lambda a,b:[f+min(d,e)for d,e,f in zip(a,a[1:],b)],t[::-1])[0]
    
    def minimumTotal_2(self, triangle):
        def combine_rows(lower_row, upper_row):
            return [upper + min(lower_left, lower_right)
                    for upper, lower_left, lower_right in
                    zip(upper_row, lower_row, lower_row[1:])]
        return reduce(combine_rows, triangle[::-1])[0]
    
    def minimumTotal_3(self, triangle):
        return reduce(lambda a, b:[min(a[i], a[i+1])+n for i, n in enumerate(b)], triangle[::-1])[0]
    
    def minimumTotal_4(self, triangle):
        dp = triangle[-1][:]
        for i in range(len(triangle)-2, -1, -1):
            for j in range(i+1):
                dp[j] = min(dp[j], dp[j+1]) + triangle[i][j]
        return dp[0] 

    def minimumTotal_5(self, triangle):
        return min(reduce(lambda a, b: [c+min(d,e) for c,d,e in zip(b,['']+a,a+[''])], triangle, [0])) 
    
    def minimumTotal_6(self, triangle):
        r = [0]
        for row in triangle:
            r = [row[i] + min (r[max(i-1,0)], r[min(i,len(r)-1)]) for i in range(len(row))]
        return min(r)  

def main():
    triangle = [[2],[3,4],[6,5,7],[4,1,8,3]]        #expect is 11
    obj = Solution()
    result = obj.minimumTotal(triangle)
    print("return result is :",result)
    
if __name__ =='__main__':
    main()   